episodes/4-advanced-layer-types.Rmd

```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
+```
+
+Now, let's find the best combination of hyperparameters using grid search. 
+Grid search is the simplest hyperparameter tuning strategy, 
+you just test all the combinations of a few predefined values for the hyperparameters that you want to vary.
+Note that this can take a some time to train (around 5 minutes or longer).
+```python
+import keras_tuner
+
+hp = keras_tuner.HyperParameters()
+
+def build_model(hp):
+    # Define values for hyperparameters to try out:
+    n_layers = hp.Int("n_layers", min_value=1, max_value=2, step=1)
+    dropout_rate = hp.Float("dropout_rate", min_value=0.2, max_value=0.8, step=0.3)
+    
+    model = create_nn_with_hp(dropout_rate, n_layers)
+    compile_model(model)
+    return model
+
+tuner = keras_tuner.GridSearch(build_model, objective='val_loss')
+
+tuner.search(train_images, train_labels, epochs=20,
+             validation_data=(val_images, val_labels))
+```
+```output
+Trial 6 Complete [00h 00m 46s]
+val_loss: 1.3021799325942993
+
+Best val_loss So Far: 1.2409346103668213
+Total elapsed time: 00h 03m 59s
+```
+Let's have a look at the results:
+
+```python
+tuner.results_summary()
+```
+```output
+Results summary
+Results in ./untitled_project
+Showing 10 best trials
+Objective(name="val_loss", direction="min")
+
+Trial 0004 summary
+Hyperparameters:
+n_layers: 2
+dropout_rate: 0.5
+Score: 1.2409346103668213
+
+Trial 0003 summary
+Hyperparameters:
+n_layers: 2
+dropout_rate: 0.2
+Score: 1.281008005142212
+
+Trial 0005 summary
+Hyperparameters:
+n_layers: 2
+dropout_rate: 0.8
+Score: 1.3021799325942993
+
+Trial 0002 summary
+Hyperparameters:
+n_layers: 1
+dropout_rate: 0.8
+Score: 1.3677740097045898
+
+Trial 0001 summary
+Hyperparameters:
+n_layers: 1
+dropout_rate: 0.5
+Score: 1.3880290985107422
+
+Trial 0000 summary
+Hyperparameters:
+n_layers: 1
+dropout_rate: 0.2
+Score: 1.4468265771865845
+```
+
+::: challenge
+
+## Hyperparameter tuning
+
+1: Looking at the grid search results, select all correct statements:
+
+- A. 6 different models were trained in this grid search run, because there are 6 possible combinations for the defined hyperparameter values
+- B. There are 2 different models trained, 1 for each hyperparameter that we want to change
+- C. 1 model is trained with 6 different hyperparameter combinations
+- D. The model with 1 layer and a dropout rate of 0.2 is the best model with a validation loss of 1.45
+- E. The model with 2 layers and a dropout rate of 0.5 is the best model with a validation loss of 1.24
+- F. We have found the model with the best possible combination of dropout rate and number of layers
+
+2 (Optional): Perform a grid search finding the best combination of the following hyperparameters: 2 different activation functions: 'relu', and 'tanh', and 2 different values for the kernel size: 3 and 4. Which combination works best?
+
+**Hint**: You can use `hp.Choice("name", ["value1", "value2"])` to use hyperparameters from a predefined set of possible values. 
```

Comment:
This hint seems a bit confusing without further explanation. Given that this is an optional extra exercise, it might be ok to leave it to the learner.

Reply:
Fixed in [c38b110](https://github.com/carpentries-incubator/deep-learning-intro/pull/437/commits/c38b110c0e25e16c087c1038278608e7d6617cf1)